doc/marc.1
.\" Copyright (C) 1997-1999, 2003, 2004, 2017-2018 Ossama Othman
.\"
.\" SPDX-License-Identifier: GFDL-1.3-or-later
.\"
.\" Process this file with
.\" groff -man -Tascii marc.1
.\"
.TH MARC 1 "2018-07-12" "MaRC" "User Commands"
.SH NAME
marc \- create digital map projections
.SH SYNOPSIS
.SY marc
.OP \-?V
.OP \-\-help
.OP \-\-usage
.OP \-\-version
.IR FILE .\|.\|.
.YS

.SH DESCRIPTION
.B marc
(Map Reprojections and Conversions) is a program designed to produce
maps from digital images or from computed values of incidence,
emission or phase angles, and latitudes and longitudes.  Multi-plane
maps can be created with digital image data in several planes and
cosines of incidence, emission, and phase angles, as well as latitudes
and longitudes in other planes.  Placement of data in a given plane is
completely arbitrary.  The user decides in what plane to place a given
set of data.  Created maps are stored in FITS (Flexible Image
Transport System) format.

Input
.IR FILE s
provided on the
.B marc
command line describe the body being mapped, observer viewing
geometry, source data, and map characteristics.  Multiple
.IR FILE s
may be specified, causing
.B marc
to process each of them.  See the
.B MaRC_input
manual for a description
of the
.B marc
input file format.

.B marc
supports the following options:
.TP
.B \-?, \-\-help
give this help list
.TP
.B \-\-usage
give a short usage message
.TP
.B \-V, \-\-version
print program version

.SH FILES
.TP
.I $XDG_CONFIG_HOME/marc
User configuration file, which defaults to
.I ~/.config/marc
if
.I $XDG_CONFIG_HOME
is not set.  See the
.B MaRC_input
manual for further details.

.SH BUGS
Conformal and equivalent projections have not been well tested or
verified.

Report bugs on the
.UR https://github.com/ossama-othman/MaRC/issues
MaRC issues page
.UE .


NEWS
1 February 2019 - MaRC 0.10

- Fixed a regression in quadratic roots calculations that caused some
  areas of an Orthographic projection to incorrectly have data of
  value zero, as well as other coordinate calculations.

- Blank areas of integer typed maps now contain the user-supplied
  DATA_BLANK value, if specified in the MaRC input file.  This allows
  FITS image readers to easily identify areas of integer typed maps
  that lack data.

- Corrected a problem where MaRC's Simple Cylindrical projection could
  not handle the case where lower and upper longitudes in the map were
  the same, signifying a 360 degree range, resulting in only single
  longitude being mapped across all map pixels.  MaRC now assumes the
  full 360 degree longitude range is being mapped if both the lower
  and upper longitude are the same, and issues a diagnostic message
  accordingly.

- The MU plane no longer needs to have the sub-solar point specified.
  It was never needed for emission angle calculations.

- The MU0 plane no longer needs to have the sub-observer point and
  range specified.  They were never needed for incidence angle
  computations.

- Improved MaRC input file parser error reporting.  The filename, line
  number and column containing the syntax error are now reported to
  the user.

- Corrected handling of position angles explicitly marked as clockwise,
  i.e. "CW", in MaRC input files.  They were erroneously treated as
  counter-clockwise position angles.

- The user-specific MaRC configuration file is now expected to be
  `$XDG_CONFIG_HOME/marc', which is `~/.config/marc' by default, to
  conform to the XDG Base Directory Specification.  The old MaRC
  configuration filename `~/.marc' is no longer supported.

- Added command line option support.  Try "marc --help".

- Improved logging by leveraging the fmt library as the underlying
  logging framework.  This allows for faster logging, and log level
  differentiated output.

- Corrected progress output for the Orthographic map projection so
  that completion percentage is printed up to 100%.

- MERCATOR map entries in MaRC input files may now use options common
  to all MaRC map projections.

- Fixed a mosaic averaging regression that caused weighted and
  unweighted averaging to be disabled if either was selected.

- Added support for generation of code coverage results for g++
  builds.  Try the following to generate the results:
        ./configure --enable-code-coverage
        make check-code-coverage

- Replaced all uses of the MaRC-specific "bodycentric" and
  "bodygraphic" latitude terminology with the NASA NAIF recommended
  "planetocentric" and "planetographic", respectively.

31 January 2018 - MaRC 0.9.10

- Reduced footprint of the MaRC program (`marc') and library
  (`libMaRC').  For example, including both the MaRC library and
  program, the static footprint was reduced by 28% for a 64-bit Linux
  build with g++ 7.2 when comparing total sizes returned by the `size'
  command.

- Data scaling values, i.e. BSCALE and BZERO, as well as the physical
  data unit (BUNIT) are now automatically written to the map FITS
  header when possible.  This allows FITS reading and viewing software
  to automatically provide the actual physical value to the user,
  instead of forcing the user to manually scale the map data.
  Currently this only applies to single plane maps containing one of
  the MaRC virtual images (MU, MU0, PHASE, LATITUDE or LONGITUDE).

  Maps with multiple planes currently write the scaling values as
  free-form text across several HISTORY cards in the FITS header
  instead of standalone BSCALE, BZERO and BUNIT cards.  In this case,
  data scaling of map values to physical values must generally be
  performed manually.

- Corrected automatic offset calculation for LONGITUDE map planes when
  creating integer typed maps.  Longitude values closer to 360 degrees
  are no longer cutoff.  Be aware that an offset must be applied to
  convert from the map longitude value to the physical longitude
  value.  See the generated map FITS header to obtain the offset
  (BZERO) value.

- Further expanded the MaRC test suite (try "make check").

- The MU, MU0 and PHASE virtual images may now contain all cosine
  values i.e. the interval [-1, 1], and are no longer limited to
  positive values alone.  There was no real need to drop the negative
  values.  To obtain the old behavior where only cosine values [0, 1]
  were mapped set the DATA_MIN keyword for the MU, MU0 and PHASE map
  plane entries in your MaRC input file to zero, e.g.:

      PLANE:
              DATA_MIN: 0
              MU: ...

- The PHASE virtual image no longer ties the existence of a datum to a
  positive cosine of the emission of the angle (mu), meaning cosine of
  phase angle data will also be plotted in areas of the map where
  mu < 0.

- Improved speed of MU, MU0 and PHASE virtual image generation by up
  to 100% by dropping a redundant data "visibility" check.

- Slightly improved overall MaRC performance by leveraging link-time
  optimization on some platforms.

- Corrected problem in conformal (Mercator and Polar Stereographic)
  maps that prevented calculation of latitudes for the entire map from
  completing in some cases.

- Dropped references to "transverse" in documentation related to the
  Mercator projection.  MaRC's Mercator projection is the regular
  variant not the transverse one.

- Improved accuracy of conformal maps through root finding
  improvements.

- Fixed memory violations in conformal maps.

20 August 2017 - MaRC 0.9.9

- Data generated by the MU, MU0, PHASE, LATITUDE and LONGITUDE virtual
  images are now automatically offset and scaled by the largest power
  of 10 suitable for maximizing the number of significant digits retained
  when storing their floating point values in integer typed maps.  For
  example, MU image data, which has a range of [-1, 1], will be scaled
  and offset to a range of [0, 200] for a map with BYTE (unsigned 8

src/marc.cpp
/**
 * @file marc.cpp
 *
 * Copyright (C) 1996-1999, 2004, 2017-2018  Ossama Othman
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
 *
 * @author Ossama Othman
 */

#include "command_line.h"
#include "parse_scan.h"
#include "strerror.h"
#include "parse.hh"
#include "lexer.hh"
#include "MapCommand.h"

#include <marc/config.h>
#include <marc/Log.h>
#include <marc/utility.h>

#include <string>
#include <cstdio>
#include <cerrno>
#include <cstdlib>


namespace MaRC
{
    /**
     * @brief @c FILE stream close functor.
     *
     * This functor closes a @c FILE stream.  It is meant for use as
     * the @c std::unique_ptr @c Deleter template parameter.
     */
    struct FILE_closer
    {
        /**
         * @param[in] stream Pointer to @c FILE stream.
         */
        void operator()(std::FILE * stream) const
        {
            std::fclose(stream);
        }
    };

    /**
     * Type alias of @c std::unique_ptr that automatically closes
     * the a @c FILE stream up on exiting the scope in which an
     * instance of this @c file_unique_ptr resides.
     */
    using FILE_unique_ptr = std::unique_ptr<std::FILE, FILE_closer>;

    // --------------------------------------------------------------

    /**
     * @brief @c Flex reentrant scanner destroyer.
     *
     * This functor destroys a reentrant Flex scanner.  It is meant
     * for use as the @c std::unique_ptr @c Deleter template
     * parameter.
     */
    struct scanner_destroyer
    {
        /**
         * @param[in] scanner Pointer to reentrant scanner object.
         */
        void operator()(yyscan_t * scanner) const
        {
            yylex_destroy(*scanner);
        }
    };

    /**
     * Type alias of @c std::unique_ptr that automatically destroys
     * a reentrant Flex scanner up on exiting the scope in which an
     * instance of this @c scanner_unique_ptr resides.
     */
    using scanner_unique_ptr =
        std::unique_ptr<yyscan_t, scanner_destroyer>;

    // --------------------------------------------------------------

    /**
     * @brief Parse a %MaRC configuration or input file.
     *
     * @param[in]     filename %MaRC configuration or input file to
     *                         parse.
     * @param[in,out] pp       %MaRC configuration parameters.
     *
     * @retval  0 Parsing succeded.
     * @retval  1 Parsing failed due to invalid input.
     * @retval  2 Parsing failed due to memory exhaustion.
     * @retval -1 Error occurred.  Check @c errno.
     */
    int parse_file(char const * filename, MaRC::ParseParameter & pp)
    {
        errno = 0;
        MaRC::FILE_unique_ptr const file(std::fopen(filename, "r"));

        if (!file) {
#ifndef NDEBUG
            constexpr std::size_t BUFLEN = 80;
            char buf[BUFLEN] = {};

            char const * const error =
                MaRC::strerror(errno, buf, MaRC::size(buf));

            MaRC::debug("Unable to open input file '{}': {}",
                        filename,
                        error);
#endif

            return -1;
        }

        yyscan_t scanner;

        if (yylex_init(&scanner) != 0)
            return -1;

        MaRC::scanner_unique_ptr safe_scanner(&scanner);

        yyset_in(file.get(), scanner);

        // For syntax error reporting.
        pp.filename = filename;

        // Parse user defaults/MaRC initialization file.
        int const parsed = ::yyparse(scanner, pp);

        if (parsed == 0) {
            // Successful parse
            MaRC::debug("MaRC input file '{}' parsed", filename);
        }

        return parsed;
    }

    /**
     * @brief Get the %MaRC configuration filename.
     *
     * Get the %MaRC configuration filename, conforming to the XDG
     * Base Directory specification.  The %MaRC configuration file
     * will be @c ~/.config/marc by default, assuming the user hasn't
     * changed the %MaRC package name at build-time.
     *
     * @see https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html
     */
    std::string get_config_filename()
    {
        // Get XDG Base Directory Specification conforming
        // configuration file.
        std::string filename;

        char const * const config_dir = std::getenv("XDG_CONFIG_HOME");
        char const * const home_dir   = std::getenv("HOME");

        if (config_dir != nullptr) {
            filename = std::string(config_dir) + "/" PACKAGE;
        } else {
            if (home_dir == nullptr)
                return filename;  // No home directory!

            constexpr char const tmp[] = "/.config/" PACKAGE;

            filename = std::string(home_dir) + tmp;
        }

        // -----------------------------------------------------
        // Warn of existence of old configuration file.
        // -----------------------------------------------------
        if (home_dir != nullptr) {
            std::string const old_config =
                std::string(home_dir) + "/.marc";

            if (access(old_config.c_str(), F_OK) == 0) {
                MaRC::warn("old configuration file `{}' exists",
                           old_config);
                MaRC::warn("expected: `{}'.", filename);
            }
        }
        // -----------------------------------------------------

        return filename;
    }

}

/// The canonical main entry point to the %MaRC process.
int main(int argc, char *argv[])
{
    int result = 0;

    try {
        MaRC::ParseParameter parse_parameter;

        std::string const config_file(MaRC::get_config_filename());

        // Parse user configuration file, if it exists.
        if (MaRC::parse_file(config_file.c_str(), parse_parameter) != 0
            && errno != ENOENT) {
            // Scanner or parser error.
            return -1;
        }

        /*
          Parse command line options, overriding corresponding
          configuration file parameters if any exist.
        */
        MaRC::command_line cl;

        if (!cl.parse(argc, argv))
            return 1;

        // Parse MaRC input files given on command line.
        for (auto const filename : cl.files())
            if (MaRC::parse_file(filename, parse_parameter) != 0)
                return -1;

        // Create the map(s).
        auto const & commands =
            parse_parameter.commands();

        for (auto & p : commands) {
            if (p->execute() != 0) {
                MaRC::error("problem during creation of map '{}'",
                            p->filename());
                result = 1;
            }
        }
    } catch (std::exception const & e) {
        /*
          Fall back on fprintf() to avoid potentially throwing another
          exception through the underlying C++ logging framework.
        */
        std::fprintf(stderr,
                     "[" PACKAGE "][error] %s\n",
                     e.what());

        return -1;
    }

    return result;
}
